### PR TITLE
fix issue 108

### PR DESCRIPTION
**Pull request by shenli**
Fix issue: https://github.com/pingcap/tidb/issues/108
Refactor update.go and support multiple-table syntax.

I change update syntax according to https://dev.mysql.com/doc/refman/5.7/en/update.html. There should be a "SET" in update statement. So some of tidb_test should be changed.

### CODE REVIEW

File: parser/parser.y [hunk 1]
```diff
@@ -656,9 +659,25 @@ Symbol:
 
 /*******************************************************************************************/
 Assignment:
-	ColumnName eq Expression
+	QualifiedIdent eq Expression
```
Comment (siddontang) [inline parser/parser.y]: QualifiedIdent contains `db.table.columnname` or `*`, you should handle these for error, maybe.

Comment (shenli) [inline parser/parser.y]: I add some checks. Maybe we will support db.tbl.col in the future.

Comment (siddontang) [inline parser/parser.y]: I think so.

File: expression/expressions/assign.go [hunk 2]
```diff
@@ -21,6 +21,7 @@ import (
 
 // Assignment is the expression for assignment, like a = 1.
 type Assignment struct {
+	TableName string
```
Comment (qiuyesuifeng) [inline expression/expressions/assign.go]: Adding one line comment is much better.

File: stmt/stmts/update.go [hunk 3]
```diff
@@ -204,103 +207,109 @@ func updateRecord(ctx context.Context, h int64, data []interface{}, t table.Tabl
...
+			}
+			// Update row
+			id, uerr := util.DecodeHandleFromRowKey(k)
+			if uerr != nil {
+				return false, errors.Trace(uerr)
+			}
+			end := start + len(tbl.Cols())
+			data := in[start:end]
+			start = end
+			tcols, uerr := getUpdateColumns(tbl, s.List, s.MultipleTable)
```
Comment (ngaut) [inline stmt/stmts/update.go]: s/uerr/err1

File: parser/parser.y [hunk 4]
```diff
@@ -656,9 +659,33 @@ Symbol:
 
 /*******************************************************************************************/
 Assignment:
-	ColumnName eq Expression
-	{
-		$$ = expressions.Assignment{ColName: $1.(string), Expr: expressions.Expr($3)}
+	QualifiedIdent eq Expression
+	{
+		name := $1.(string)
+		fs := strings.Split(name, ".")
```
Comment (ngaut) [inline parser/parser.y]: What does fs mean ?

File: parser/parser.y [hunk 5]
```diff
@@ -656,9 +659,33 @@ Symbol:
...
+		var (
+			tn string
```
Comment (ngaut) [inline parser/parser.y]: s/tn/tableName

File: parser/parser.y [hunk 6]
```diff
@@ -656,9 +659,33 @@ Symbol:
...
+			cn string
```
Comment (ngaut) [inline parser/parser.y]: s/cn/colName

File: stmt/stmts/update.go [hunk 7]
```diff
@@ -204,103 +207,109 @@ func updateRecord(ctx context.Context, h int64, data []interface{}, t table.Tabl
...
+		for _, entry := range rowKeys.Keys {
+			tbl := entry.Tbl
+			k := entry.Key
+			_, ok := updatedRowKeys[k]
+			if ok {
+				// Each matching row is updated once, even if it matches the conditions multiple times.
+				continue
+			}
+			// Update row
+			id, uerr := util.DecodeHandleFromRowKey(k)
```
Comment (ngaut) [inline stmt/stmts/update.go]: Does id means handle ?

File: parser/parser.y [hunk 8]
```diff
@@ -656,9 +659,33 @@ Symbol:
 
 /*******************************************************************************************/
 Assignment:
-	ColumnName eq Expression
-	{
-		$$ = expressions.Assignment{ColName: $1.(string), Expr: expressions.Expr($3)}
+	QualifiedIdent eq Expression
```
Comment (qiuyesuifeng) [inline parser/parser.y]: Suggest moving qualified assignment logic into `expressions/assign.go` which can make parser cleaner.

File: parser/parser.y [hunk 9]
```diff
@@ -656,9 +659,33 @@ Symbol:
...
+	{
```
Comment (coocood) [inline parser/parser.y]: We can define an item like `SpecificIdent` that does not contain wild card.

File: stmt/stmts/update.go (hunk 7 above)
Comment (shenli) [inline stmt/stmts/update.go]: yes

File: parser/parser.y (hunk 8 above)
Comment (shenli) [inline parser/parser.y]: ok

File: parser/parser.y (hunk 9 above)
Comment (shenli) [inline parser/parser.y]: ok

File: parser/parser.y (hunk 4 above)
Comment (shenli) [inline parser/parser.y]: split a string to fields

File: stmt/stmts/update.go (hunk 7 above)
Comment (ngaut) [inline stmt/stmts/update.go]: So rename id to handle maybe better.

File: parser/parser.y (hunk 4 above)
Comment (ngaut) [inline parser/parser.y]: strs maybe a better name.

File: stmt/stmts/update.go [hunk 10]
```diff
@@ -82,37 +85,27 @@ func (s *UpdateStmt) SetText(text string) {
...
 }
 
-func getUpdateColumns(t table.Table, assignList []expressions.Assignment) ([]*column.Col, error) {
-	tcols := make([]*column.Col, len(assignList))
-	for i, asgn := range assignList {
+func getUpdateColumns(t table.Table, assignList []expressions.Assignment, multipleTable bool) ([]*column.Col, error) {
+	tcols := make([]*column.Col, 0, len(assignList))
+	tname := t.TableName()
+	for _, asgn := range assignList {
+		if multipleTable {
```
Comment (qiuyesuifeng) [inline stmt/stmts/update.go]: For table `a(c1 int, c3 int), b(c2 int)`
then `update a, b set a.c1 = 1 and b.c2 = 2 and c3 = 3`,
can this be supported?

File: stmt/stmts/update.go [hunk 11]
```diff
@@ -82,37 +85,27 @@ func (s *UpdateStmt) SetText(text string) {
...
+			if !strings.EqualFold(tname.O, asgn.TableName) {
+				continue
+			}
+		}
 		col := column.FindCol(t.Cols(), asgn.ColName)
 		if col == nil {
+			if multipleTable {
```
Comment (qiuyesuifeng) [inline stmt/stmts/update.go]: In my impression, we only check columns validity in `select fields`, `where`, `groupby`, `having` and `orderby`, perhaps not `join`, i am not sure, it is better to have a try.

File: stmt/stmts/update.go [hunk 12]
```diff
@@ -122,7 +115,16 @@ func getInsertValue(name string, cols []*column.Col, row []interface{}) (interfa
 	return nil, errors.Errorf("unknown field %s", name)
 }
 
-func updateRecord(ctx context.Context, h int64, data []interface{}, t table.Table, tcols []*column.Col, assignList []expressions.Assignment, insertData []interface{}) error {
+func getIdentValue(name string, fields []*field.ResultField, row []interface{}, flag uint32) (interface{}, error) {
```
Comment (qiuyesuifeng) [inline stmt/stmts/update.go]: I remember that in plans, we have a function called `getIdentValue`, can use that one?

File: stmt/stmts/update_test.go [hunk 13]
```diff
@@ -86,3 +86,46 @@ func (s *testStmtSuite) TestUpdate(c *C) {
...
+		)
+		rows.Scan(&id, &price)
+		if id == 11 {
+			c.Assert(price, Equals, "month_price_11")
+		} else if id == 12 {
+			c.Assert(price, Equals, "items_price_12")
+		} else if id == 13 {
+			c.Assert(price, Equals, "month_price_13")
+		}
+	}
```
Comment (qiuyesuifeng) [inline stmt/stmts/update_test.go]: Need mustCommit?

File: stmt/stmts/update.go (hunk 12 above)
Comment (shenli) [inline stmt/stmts/update.go]: It is in plans. I will export it.

File: stmt/stmts/update_test.go (hunk 13 above)
Comment (siddontang) [inline stmt/stmts/update_test.go]: using a table for expected id -> price may be better.

**Comment by shenli** [comment]
1. Merge with master.
2. Replace Do with Next in update.go 
3. Address comments

PTAL

File: plan/plans/join.go [hunk 14]
```diff
@@ -330,21 +330,14 @@ func appendRow(prefix []interface{}, in []interface{}) []interface{} {
...
 		}
 	}
-	if rks == nil {
-		rks = &RowKeyList{}
-	}
-	in = append(in, rks)
 	row := make([]interface{}, 0, len(prefix)+len(in))
 	row = append(row, prefix...)
 	row = append(row, in...)
+	row = append(row, rks)
```
Comment (siddontang) [inline plan/plans/join.go]: now, the row key is still in the end of the row? not in Row struct?

Comment (shenli) [inline plan/plans/join.go]: The Do function will be abandoned. In update.go we use Next. @coocood will remove Do() later.

File: expression/expressions/assign.go [hunk 15]
```diff
@@ -29,5 +33,31 @@ type Assignment struct {
...
+		colName string
+	)
+	if len(strs) == 1 {
+		colName = strs[0]
+	} else if len(strs) == 2 {
+		tblName = strs[0]
+		colName = strs[1]
+	} else {
+		// TODO: we should support db.tbl.col later.
+		return nil, errors.Errorf("Wrong format of Assignment key: %s", key)
```
Comment (qiuyesuifeng) [inline expression/expressions/assign.go]: Invalid format or bad format is better.

**Comment by ngaut** [comment]
LGTM

**Pull request merged**